### PR TITLE
🎮 Added support for terrains without OTC

### LINKED ISSUE #1841: Compatibility with terrains with missing otc file
Try to load a terrain with missing otc file:
```
[RoR|Terrain] Error reading main *.otc file, message: 'FileNotFoundException: Cannot locate resource Houston2013.otc in resource group General. in ResourceGroupManager::openResource at /home/babis/Downloads/ror-dependencies/Source/ogre/OgreMain/src/OgreResourceGroupManager.cpp (line 698)', type: <Ogre::Exception>.
```
Example terrain: http://www.mediafire.com/file/nn36bdj9bo99khv/Traxxas_Stadium_Freestle_4.zip/file

It would be nice if this will work again, like older RoR versions (pre 0.4.7)

### PR DESCRIPTION
**Pull request by AnotherFoxGuy**
Closes #1841

### CODE REVIEW

**Comment by ulteq** [comment]
I'm not a fan of the automatic code reformat tool that adds spaces between method names and braces.

**Comment by ulteq** [comment]
`Traxxas_Stadium_Freestle_4.zip` segfaults with:
```
 ===== LOADING TERRAIN GEOMETRY TraxxasMOFreeS.terrn2
  ## Loading Terrain Geometry
[RoR|Terrain] Error reading main *.otc file, message: 'FileNotFoundException: Cannot locate resource Houston2013.otc in resource group General. in ResourceGroupManager::openResource at /home/ulteq/ror-dependencies/Source/ogre/OgreMain/src/OgreResourceGroupManager.cpp (line 704)', type: <Ogre::Exception>.

Thread 1 "RoR" received signal SIGSEGV, Segmentation fault.
TerrainGeometryManager::SetupLayers (this=<optimized out>, page=..., terrain=0x0) at /home/ulteq/rigs-of-rods/source/main/terrain/TerrainGeometryManager.cpp:540
540         if (!terrain)
(gdb) bt
#0  TerrainGeometryManager::SetupLayers (this=<optimized out>, page=..., terrain=0x0) at /home/ulteq/rigs-of-rods/source/main/terrain/TerrainGeometryManager.cpp:540
#1  0x0000555555904e98 in TerrainGeometryManager::configureTerrainDefaults (this=0x55555822c830) at /usr/include/c++/8.2.1/ext/aligned_buffer.h:70
#2  0x00005555559058b3 in TerrainGeometryManager::InitTerrain (this=this@entry=0x55555822c830, otc_filename="Houston2013.otc") at /home/ulteq/rigs-of-rods/source/main/terrain/TerrainGeometryManager.cpp:341
#3  0x00005555559098d2 in TerrainManager::LoadAndPrepareTerrain (this=this@entry=0x555559669eb0, filename="TraxxasMOFreeS.terrn2") at /usr/include/c++/8.2.1/bits/basic_string.h:927
#4  0x0000555555647b83 in SimController::LoadTerrain (this=0x7fffffffdf30) at /usr/include/c++/8.2.1/bits/basic_string.h:927
#5  0x0000555555648f90 in SimController::SetupGameplayLoop (this=this@entry=0x7fffffffdf30) at /home/ulteq/rigs-of-rods/source/main/gameplay/RoRFrameListener.cpp:2056
#6  0x0000555555606c99 in main (argc=<optimized out>, argv=<optimized out>) at /home/ulteq/rigs-of-rods/source/main/main.cpp:319
```

**Comment by ulteq** [comment]
My suggested solution would look like this: https://github.com/ulteq/rigs-of-rods/commit/60a34546460b3b929e2078dd799094a074787b7c

**Comment by ulteq** [comment]
The `OTCFileformat` class looks perfect now.

But I wouldn't change anything in the `TerrainGeometryManager` class other than: https://github.com/ulteq/rigs-of-rods/commit/60a34546460b3b929e2078dd799094a074787b7c#diff-6a837bc7d5e0a57075af0073b134f2f9

**Comment by ulteq** [comment]
@AnotherFoxGuy Why did you revert your `OTCFile` initializer changes?

File: source/main/resources/terrn2_fileformat/Terrn2Fileformat.cpp [hunk 1]
```diff
@@ -45,14 +45,7 @@ bool Terrn2Parser::LoadTerrn2(Terrn2Def& def, Ogre::DataStreamPtr &ds)
         return false;
     }
 
-    def.ogre_ter_conf_filename = file.GetStringEx("GeometryConfig", "General");
-    // otc = ogre terrain config
-    if (def.ogre_ter_conf_filename.find(".otc") == String::npos)
```
Comment (ulteq) [inline source/main/resources/terrn2_fileformat/Terrn2Fileformat.cpp]: Did the old parser ignore the file type?

File: source/main/resources/otc_fileformat/OTCFileformat.cpp [hunk 2]
```diff
@@ -30,6 +30,11 @@
 #include <algorithm>
 #include <OgreException.h>
 
+RoR::OTCParser::OTCParser()
+{
+    m_def = std::make_shared<RoR::OTCFile>();
```
Comment (ulteq) [inline source/main/resources/otc_fileformat/OTCFileformat.cpp]: We don't need this anymore, right?

File: source/main/resources/otc_fileformat/OTCFileformat.h [hunk 3]
```diff
@@ -93,6 +93,8 @@ struct OTCFile
 class OTCParser
 {
 public:
+    OTCParser();
```
Comment (ulteq) [inline source/main/resources/otc_fileformat/OTCFileformat.h]: Do we still need this?

File: source/main/resources/otc_fileformat/OTCFileformat.cpp (hunk 2 above)
Comment (only-a-ptr) [reply source/main/resources/otc_fileformat/OTCFileformat.cpp]: This is a mess, the object is created on 2 places: https://github.com/RigsOfRods/rigs-of-rods/pull/1849/files/4d600bc52ad894dd873cca59a64327260e39ba91#diff-fb9af0b3e9573c633daff2a2a4cba402R41

File: source/main/terrain/TerrainGeometryManager.cpp [hunk 4]
```diff
@@ -137,6 +137,12 @@ TerrainGeometryManager::TerrainGeometryManager(TerrainManager* terrainManager)
     , m_was_new_geometry_generated(false)
     , terrainManager(terrainManager)
 {
+    m_spec = std::make_shared<RoR::OTCFile>();
```
Comment (only-a-ptr) [inline source/main/terrain/TerrainGeometryManager.cpp]: This is dirty, this object should come from OTCParser as a factory.

**Comment by ulteq** [comment]
We had a very simple solution earlier today ... but I can't get it together anymore. :angry:

**Comment by ulteq** [comment]
> What was wrong with #1849?

My main problem with this is that I can't keep track of the effects of all the changes.

**Pull request closed without merging**